libvast/src/data.cpp

```diff
@@ -285,4 +338,108 @@ bool convert(const data& d, json& j, const type& t) {
   return convert(d, j);
 }
 
+namespace {
+
+data parse(const YAML::Node& node) {
+  switch (node.Type()) {
+    case YAML::NodeType::Undefined:
+    case YAML::NodeType::Null:
+      return data{};
+    case YAML::NodeType::Scalar: {
+      auto str = node.as<std::string>();
+      data result;
+      // Attempt some type inference.
+      if (parsers::boolean(str, result))
+        return result;
+      // Attempt maximum type inference.
+      if (parsers::data(str, result))
+        return result;
+      // Take the input as-is if nothing worked.
+      return str;
+    }
+    case YAML::NodeType::Sequence: {
+      list xs;
+      xs.reserve(node.size());
+      for (size_t i = 0; i < node.size(); ++i)
+        xs.push_back(parse(node[i]));
+      return xs;
+    }
+    case YAML::NodeType::Map: {
+      record xs;
+      xs.reserve(node.size());
+      for (auto& pair : node)
+        xs.emplace(pair.first.as<std::string>(), parse(pair.second));
+      return xs;
+    }
+  }
+  return {}; // makes GCC happy
```

Comment:
Wouldn't `die(...)` be more appropriate here?

Reply:
Good call.